pennylane/capture/explanations.md

```diff
@@ -0,0 +1,291 @@
+
+    creating an instance of type <class '__main__.MyClass'> with (0.1,), {'a': 2}. 
+    now creating an instance in __init__
+
+
+
+
+
+    <__main__.MyClass at 0x11c5a2810>
+
+
+
+So far, we've just added print statements around default behavior.  Let's try something more radical
+
+
+```python
+class MetaClass2(type):
+
+    def __call__(cls, *args, **kwargs):
+        return 2.0
+
+class WeirdClass(metaclass=MetaClass2):
+
+    def __init__(self, *args, **kwargs):
+        print("Am I here?")
+        self.args = args
+```
+
+You can see now that instead of actually getting an instance of `WeirdClass`, we just get `2.0`. 
+
+Using a metaclass, we can hijack what happens when a type is called.
+
+
+```python
+out = WeirdClass(1.0)
+out, out == 2.0
+```
+
+
+
+
+    (2.0, True)
+
+
+
+## Putting Primitives and Metaprogramming together
+
+We have two goals that we need to accomplish with our meta class.
+
+1. Create an associated primitive every time we define a new class type
+2. Hijack creating a new instance to use `primitive.bind` instead
+
+
+```python
+class PrimitiveMeta(type):
+
+    def __init__(cls, *args, **kwargs):
+        # here we set up the primitive
+        primitive = jax.core.Primitive(cls.__name__)
+
+        @primitive.def_impl
+        def _(*inner_args, **inner_kwargs):
+            # just normal class creation if not tracing
+            return type.__call__(cls, *inner_args, **inner_kwargs)
+
+        @primitive.def_abstract_eval
+        def _(*inner_args, **inner_kwargs):
+            # here we say that we just return an array of type float32 and shape (1,)
+            # other abstract types could be used instead
+            return jax.core.ShapedArray((1,), jax.numpy.float32)
+
+        cls._primitive = primitive
+
+    def __call__(cls, *args, **kwargs):
+        return cls._primitive.bind(*args, **kwargs)
+```
+
+
+```python
+class PrimitiveClass(metaclass=PrimitiveMeta):
+
+    def __init__(self, a):
+        self.a = a
+
+    def __repr__(self):
+        return f"PrimitiveClass({self.a})"
+```
+
+What happens if we just create a class normally as is?
+
+
+```python
+PrimitiveClass(1.0)
+```
+
+
+
+
+    PrimitiveClass(1.0)
+
+
+
+But now it can also be used in tracing as well
+
+
+```python
+jax.make_jaxpr(PrimitiveClass)(1.0)
+```
+
+
+
+
+    { lambda ; a:f32[]. let b:f32[1] = PrimitiveClass a in (b,) }
+
+
+
+Great!👍
+
+Now you can see that the problem is that we lied in our definition of abstract evaluation.  Jax thinks that `PrimitiveClass` returns something of shape `(1,)` and type `float32`.
+
+But jax doesn't have an abstract type that really describes "PrimitiveClass".  So we need to define an register our own.
+
+
+```python
+class AbstractPrimitiveClass(jax.core.AbstractValue):
+
+    def __eq__(self, other):
+        return isinstance(other, AbstractPrimitiveClass)
+
+    def __hash__(self):
+        return hash("AbstractPrimitiveClass")
+
+# not quite sure what it does, but we have to do this...
```

Comment:
We don't need to elaborate, but I _think_ I understand why we need this:
The function `raise_to_shaped` guarantees to take abstract or concrete values and return suitable shaped, abstract objects. This way, abstract and concrete values can be handled on a common ground.
As there is no default implementation, but `raise_to_shaped` essentially relies on `raise_to_shaped_mappings` to contain a proper conversion method for each abstract type, we need to add one for each type we introduce.

Reply:
Still don't really understand it, but removing the comment.